File: chapter_4_trees_graphs/js/4.8-firstCommonAncestor.js
const { TreeNode } = require('./tree')

/* LINK TO PARENTS EXIST */

/* Finds the first common ancestor of two nodes in a binary tree
 * if there are links to parents */
const firstCommonAncestor = (first, second) => {
  const delta = findDepth(first) - findDepth(second)
  long = delta > 0 ? first : second
  short = delta > 0 ? second : first
  new_long = moveUpBy(long, Math.abs(delta))
  while (new_long && short && new_long != short) {
    new_long = new_long.parent
    short = short.parent
  }
  return !new_long || !short ? null : short
}

/* Moves a node up by a given number of nodes */
const moveUpBy = (node, delta) => {
  while (delta && node) {
    node = node.parent
    delta--
  }
  return node
}

/* Finds the depth from a node of a binary tree */
const findDepth = node => {
  depth = 0
  while (node) {
    node = node.parent
    depth++
  }
  return depth
}

/* LINK TO PARENTS DOES NOT EXIST */

/* Finds the first common ancestor of two nodes in a binary tree
 * if there are no links to parents */
const commonAncestor = (root, first, second) => {
  if (!covers(root, first) || !covers(root, second)) return null
  return commonAncestorHelper(root, first, second)
}

/* Helper */
const commonAncestorHelper = (root, first, second) => {
  if (!root || root === first || root === second) return root
  is_first_left = covers(root.left, first)
  is_second_left = covers(root.left, second)
  if (is_first_left != is_second_left) return root
  const childSide = is_first_left ? root.left : root.right
  return commonAncestorHelper(childSide, first, second)
}

/* Determines if a node has another node in subtree */
const covers = (root, node) => {
  if (!root) return false
  if (root === node) return true
  return covers(root.left, node) || covers(root.right, node)
}

const n1 = new TreeNode(1)
const n2 = n1.addLeft(2)
const n3 = n1.addRight(3)
const n4 = n2.addLeft(4)
const n5 = n2.addRight(5)
const n6 = n3.addLeft(6)
const n7 = n3.addRight(7)
const n8 = n4.addLeft(8)
const n9 = n4.addRight(9)
const n10 = n5.addLeft(10)
const n11 = n5.addRight(11)
const n12 = n6.addLeft(12)
const n13 = n6.addRight(13)
const n14 = n7.addLeft(14)
const n15 = n7.addRight(15)

console.log(firstCommonAncestor(n15, n8).value) // 1
console.log(firstCommonAncestor(n11, n9).value) // 2
console.log(firstCommonAncestor(n12, n13).value) // 6

console.log(commonAncestor(n1, n15, n8).value) // 1
console.log(commonAncestor(n1, n11, n9).value) // 2
console.log(commonAncestor(n1, n12, n13).value) // 6

File: chapter_4_trees_graphs/python/4.8-first_common_ancestor.py
TreeNode = __import__('tree').TreeNode

# LINK TO PARENTS EXIST


def first_common_ancestor(first_node, second_node):
    """
    Finds the first common ancestor of two nodes in a binary tree
    if there are links to parents
    """
    delta = find_depth(first_node) - find_depth(second_node)
    if delta > 0:
        longer = first_node
        shorter = second_node
    else:
        longer = second_node
        shorter = first_node
    n_longer = move_up_by(longer, abs(delta))
    while shorter and n_longer and shorter != n_longer:
        shorter = shorter.parent
        n_longer = n_longer.parent
    if shorter is None or n_longer is None:
        return None
    return shorter


def move_up_by(node, delta):
    """
    Moves a node up by a given number of nodes
    """
    curr = node
    while delta > 0 and curr:
        curr = curr.parent
        delta -= 1
    return curr


def find_depth(node):
    """
    Finds the depth from a node of a binary tree
    """
    depth = 0
    while node:
        node = node.parent
        depth += 1
    return depth

# LINK TO PARENTS DOES NOT EXIST


def common_ancestor(root, first, second):
    """
    Finds the first common ancestor of two nodes in a binary tree
    if there are no links to parents
    """
    # Error check - if one node is not in the tree
    if not covers(root, first) or not covers(root, second):
        return None
    return common_ancestor_helper(root, first, second)


def common_ancestor_helper(root, first, second):
    """
    Helper
    """
    if not root or first == root or second == root:
        return root
    first_on_left = covers(root.left, first)
    second_on_left = covers(root.left, second)
    if first_on_left != second_on_left:
        return root
    side = root.left if first_on_left else root.right
    return common_ancestor_helper(side, first, second)


def covers(root, node):
    """
    Determines if a node has another node in subtree
    """
    if not root:
        return False
    if root == node:
        return True
    return covers(root.left, node) or covers(root.right, node)


n1 = TreeNode(1)
n2 = n1.add_left(2)
n3 = n1.add_right(3)
n4 = n2.add_left(4)
n5 = n2.add_right(5)
n6 = n3.add_left(6)
n7 = n3.add_right(7)
n8 = n4.add_left(8)
n9 = n4.add_right(9)
n10 = n5.add_left(10)
n11 = n5.add_right(11)
n12 = n6.add_left(12)
n13 = n6.add_right(13)

n15 = TreeNode(15)

print(first_common_ancestor(n9, n5))  # n2
print(first_common_ancestor(n9, n15))  # None

print(common_ancestor(n1, n9, n5))  # n2
print(common_ancestor(n1, n9, n15))  # None
